zetaclient/chains/evm/signer/signer.go

```diff
@@ -201,11 +205,46 @@ func (signer *Signer) Sign(
 	return signedTX, sig[:], hashBytes[:], nil
 }
 
-// Broadcast takes in signed tx, broadcast to external chain node
-func (signer *Signer) Broadcast(tx *ethtypes.Transaction) error {
-	ctxt, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+func newTx(
+	chainID *big.Int,
+	data []byte,
+	to ethcommon.Address,
+	amount *big.Int,
+	gas Gas,
+	nonce uint64,
+) (*ethtypes.Transaction, error) {
+	if err := gas.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid gas parameters")
+	}
+
+	if gas.isLegacy() {
+		return ethtypes.NewTx(&ethtypes.LegacyTx{
+			To:       &to,
+			Value:    amount,
+			Data:     data,
+			GasPrice: gas.Price,
+			Gas:      gas.Limit,
+			Nonce:    nonce,
+		}), nil
+	}
+
+	return ethtypes.NewTx(&ethtypes.DynamicFeeTx{
+		ChainID:   chainID,
+		To:        &to,
+		Value:     amount,
+		Data:      data,
+		GasFeeCap: gas.Price,
```

Comment:
As far as I remember, the GasFeeCap (a.k.a. `maxFeePerGas`) is not a same thing as `gasPrice`. The `maxFeePerGas` is the upper limit of gas price that the sender is willing to pay. The deterministic estimation of `maxFeePerGas` could be done by `maxFeePerGas = gasTipCap + baseFee * 2`

In our case, we'll need to know the value of `baseFee` as the 1st step. This can be calculated by
`baseFee = gasPrice - gasTipCap`.  The resulted number might not match current market of the `baseFee`, but it seems fine (because Ethereum protocol doesn't require increase on `baseFee` for tx replacement).

After getting the `baseFee`, we'll then be able to calculate `maxFeePerGas` accordingly. The formulas and docs (the Alchemy one was helpful to me) are in the of [original issue](https://github.com/zeta-chain/node/issues/1952) .

Reply:
https://github.com/zeta-chain/node/pull/2634/commits/2260925ee1fd4c0b4613559d6946f59882401708